### PR TITLE
Support for assuming role before accessing KeySpaces, helps in cross account access

### PR DESCRIPTION
**Pull request by akshayar**
*Issue #, if available:*

*Description of changes:*
Support for providing IAM Role ARN is added. With this a cross account access from Spark to KeySpaces is made possible. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/test/java/software/aws/mcs/auth/TestSigV4Config.java [hunk 1]
```diff
@@ -55,7 +55,7 @@ public static void main(String[] args) throws Exception {
         //By default the reference.conf is loaded by the driver which contains all defaults.
         //You can override this by providing reference.conf on the classpath
         //to isolate test you can load conf with a custom name
-        URL url = TestSigV4Config.class.getClassLoader().getResource("keyspaces-reference.conf");
+        URL url = TestSigV4Config.class.getClassLoader().getResource("keyspaces-reference-norole.conf");
```
Comment (dchenbec) [inline src/test/java/software/aws/mcs/auth/TestSigV4Config.java:58]: I would create a new test class for this feature. This looks like maybe you checked in some test work in progress

**Review by dchenbec — CHANGES_REQUESTED** [review]
Thanks for the PR. There are a couple of changes that need to be made, please ping me if you have questions

File: src/test/java/software/aws/mcs/auth/TestSigV4Config.java (hunk 1 above)
Comment (akshayar) [reply src/test/java/software/aws/mcs/auth/TestSigV4Config.java]: Created a new class to test Assume IAM role config.

File: src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java [hunk 2]
```diff
@@ -373,4 +388,37 @@ static int indexOf(byte[] target, byte[] pattern) {
...
+                .build();
+    }
+
+    /**
+     * Gets the default region for SigV4 if region is not provided.
+     * @return Default region
+     */
+    private static String getDefaultRegion() {
+        DefaultAwsRegionProviderChain chain = new DefaultAwsRegionProviderChain();
+        return Optional.ofNullable(chain.getRegion()).orElse(Region.US_EAST_1).toString();
```
[the remark this answers is not in the document]
Comment (akshayar) [reply src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java]: It was just safe coding, did not check the doc on chain.getRegion. Updated as per your suggestion and also refactored the code you mentioned.

File: README.md [hunk 3]
```diff
@@ -138,3 +154,24 @@ The following is an example of this.
...
+    datastax-java-driver {
+        basic.load-balancing-policy {
+            class = DefaultLoadBalancingPolicy
+            local-datacenter = us-east-2
+        }
+        advanced {
+            auth-provider = {
+                class = software.aws.mcs.auth.SigV4AuthProvider
+                aws-region = us-east-2
+                aws-role = "arn:aws:iam::ACCOUNT_ID:role/ROLE_NAME"
```
[the remark this answers is not in the document]
Comment (akshayar) [reply README.md:167]: I have shared this code with customer. They may have to make changes. Need to check where they are with implementation.

File: README.md [hunk 4]
```diff
@@ -35,29 +36,44 @@ You can specify the Region using one of the following four methods:
...
+* Constructor
+* Configuration
+
+### Constructor
+
+One of the constructors for `software.aws.mcs.auth.SigV4AuthProvider` takes two String , first represening the region and second one representing IAM role ARN. 
+
+### Configuration
+
+Set the IAM Role explicitly in your `advanced.auth-provider.class` configuration (see example below), by specifying the `advanced.auth-provider.aws-role` property.
```
[the remark this answers is not in the document]
Comment (akshayar) [reply README.md]: I have shared this code with customer. They may have to make changes. Need to check where they are with implementation.

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (akshayar) [reply README.md]: Updated to aws-role-arn

File: src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java [hunk 5]
```diff
@@ -373,4 +385,42 @@ static int indexOf(byte[] target, byte[] pattern) {
...
+     * @return The STS role credential provider
+     */
+    private static StsAssumeRoleCredentialsProvider createSTSRoleCredentialProvider(String roleArn,
+                                                                     String stsRegion) {
+        //Get role name from ARN
+        String[] arnParts = roleArn.split("/");
+        if(arnParts.length < 2){
+            throw new IllegalArgumentException("Invalid role ARN");
+        }
+        String roleName = arnParts[arnParts.length - 1];
```
Comment (dchenbec) [inline src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java:403]: Would you mind splitting this out into a helper static method and adding a unit test for it (with both happy and sad paths)? Also, should we add a `@Nonnull` to the `roleArn` and `stsRegion` parameters?

File: src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java [hunk 6]
```diff
@@ -130,7 +132,8 @@ public String getPath() {
      * Unused for this plugin.
      */
     public SigV4AuthProvider(DriverContext driverContext) {
-        this(driverContext.getConfig().getDefaultProfile().getString(REGION_OPTION, null));
+        this(driverContext.getConfig().getDefaultProfile().getString(REGION_OPTION, getDefaultRegion()),
+                driverContext.getConfig().getDefaultProfile().getString(ROLE_OPTION, null));
```
Comment (dchenbec) [inline src/main/java/software/aws/mcs/auth/SigV4AuthProvider.java:136]: nit: can we align the `driverContext` on both lines please?

**Review by dchenbec — CHANGES_REQUESTED** [review]
It looks good but there is a small change to improve unit testing I think needs to be made

**Review by dchenbec — CHANGES_REQUESTED** [review]
It looks good but there is a small change to improve unit testing I think needs to be made

**Review by dchenbec — APPROVED** [review]

**Review by pranayw333 — APPROVED** [review]

**Pull request merged**